src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2414,21 +2412,39 @@ Status Session::receive_download_message(const DownloadMessage& message)
         return status;
     }
 
-    version_type server_version = m_progress.download.server_version;
+    // Start with the download server version from the last download message, since the changesets in the new
+    // download message must have a remote version greater than (or equal to for FLX) this value.
+    version_type last_remote_version = m_progress.download.server_version;
     version_type last_integrated_client_version = m_progress.download.last_integrated_client_version;
+    // If there are 2 or more changesets in this message, track whether or not the changesets all have the same
+    // download_server_version to help with determining if this is a single message server-initiated bootstrap or not.
+    bool same_remote_version = last_in_batch && message.changesets.size() > 1;
```

Comment:
if last_in_batch is false, then same_remote_version is false too and its value never changes. This is incorrect for multi-message bootstraps (where you rely on batch_state). I think this value should always reflect the reality.

On another note, computing same_remote_version seems over-complicated. I think something like this should work:
`bool bootstrap_in_progress = !last_in_batch || last_remote_version == progress.download.server_version || (message.changesets.size() > 1 && message.changesets.front().remote_version == message.changesets.back().remote_version)`

where `last_remote_version = progress.download.server_version` once the message is processed (and initialized with m_progress.download.server_version).
You could add the check for query version too and get rid off `is_steady_state_download_message` completely.